url10fb6bd43e_guwere_P2PSearch_tgz-pJ8-MainJ8/src/Controller.java

```diff
@@ -0,0 +1,577 @@
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.net.DatagramPacket;
+import java.net.InetSocketAddress;
+import java.net.SocketException;
+import java.util.*;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+
+public final class Controller implements Runnable {
+
+    private int nodeID;
+    private BlockingQueue receiveQueue;// queues allow for complete decoupling of receiver<->controller<->sender
+    private BlockingQueue sendQueue;
+    private InetSocketAddress bootstrapNode;
+    private Map<Integer,Contact> routingTable;
+    private Indexer indexer;
+    //timers---
+    private Timer rTableTimer;
+    private Waiter searchResponseWaiter;
+    private Map<Integer,Timer> pingTimers;
+    private Map<String,Timer> indexTimers;
+    //----
+    private Vector<SearchResponse> resultResponses;//for each search invocation by the user the search responses are buffered before sent back
+    public Controller(BlockingQueue rQueue, BlockingQueue sQueue){
+         receiveQueue = rQueue;
+         sendQueue = sQueue;
+         rTableTimer = new Timer();
+        //start the routing table cleanup procedure after refreshTime on every refreshTime interval to weed out stale contacts
+         rTableTimer.scheduleAtFixedRate(new RefreshRoutingTable(),NetworkProperties.refreshTime*1000,NetworkProperties.refreshTime*1000);
+         routingTable = new ConcurrentHashMap<Integer, Contact>();//access from multiple threads
+         pingTimers = new ConcurrentHashMap<Integer, Timer>();
+        indexTimers = new ConcurrentHashMap<String, Timer>();
+
+    }
+
+    @Override
+    public void run() {
+        while(true){
+            try {
+            	DatagramPacket receivedPacket =(DatagramPacket)receiveQueue.take();
+            	//the queue blocks if empty
+                handleReceivedPacket(receivedPacket);
+            } catch (InterruptedException e) {
+                System.err.println("Could not take from receive queue in Controller");
+                e.printStackTrace();
+            }
+        }
+    }
+    private static int getAsInteger(JSONObject jsonObj,String fieldName){
+        int result = 0;
+        if(jsonObj.containsKey(fieldName)){
+            try{//there are some wierd conversion going on when constructing the json
+                result = Integer.parseInt((String)jsonObj.get(fieldName));
+            }catch(NumberFormatException nfe){}
+        }
+        return result;
+    }
+    @SuppressWarnings("unchecked")
+    private void handleReceivedPacket(DatagramPacket packet){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 123 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.